libnmstate/nm/device.py

```diff
@@ -64,18 +64,28 @@ def deactivate(dev):
 
     For software devices, deactivation removes the devices from the kernel.
     """
-    client = nmclient.client()
+    mainloop = nmclient.mainloop()
+    mainloop.push_action(_safe_deactivate_async, dev)
+
+
+def _safe_deactivate_async(dev):
     act_connection = dev.get_active_connection()
-    if act_connection:
-        mainloop = nmclient.mainloop()
-        user_data = mainloop
-        mainloop.push_action(
-            client.deactivate_connection_async,
-            act_connection,
-            mainloop.cancellable,
-            _deactivate_connection_callback,
-            user_data,
-        )
+    mainloop = nmclient.mainloop()
+    if not act_connection or act_connection.props.state in (
+            nmclient.NM.ActiveConnectionState.DEACTIVATING,
+            nmclient.NM.ActiveConnectionState.DEACTIVATED):
+        # No callback is expected, so we should call the next one.
```

Comment:
The comment is there to explain why we are now triggering the next action in the mainloop (usually we do it in the callback).
It goes like this:
Usually we collect the actions we intend to issue, order them in sequence and when the mainloop starts we fire the first. The next action will be fired only when the previous one finished (i.e. the callback returned and is successful).
In this case, we want to wrap the deactivation action with a "safe" version of it.
Now, if there is nothing to deactivate, we do nothing and just call the next action. But if we do need to do something, we delay the next action for the callback.

The reason we do not do it outside the mainloop is because we cannot predict the future. It does not matter what is the state when this is checked, it matters what happens very close to the time we actually run the action.